### PR TITLE
Fix lazy decomposition: incremental SVD, masks, reproject, NMF, lazy pipeline, parity fixes

### PR DESCRIPTION
**Pull request by francisco-dlp**
## Summary

This PR fixes several pre-existing bugs in `LazySignal.decomposition()`, brings its feature set much closer to parity with the non-lazy `MVA.decomposition()`, adds a fully lazy decompose → reconstruct → save pipeline, and deprecates the old bespoke `ORPCA`/`ORNMF` class API in favour of a scikit-learn-compatible one.

---

## Bug fixes

- **#3607** — Poissonian noise normalisation was silently a no-op: `coeff.map_blocks(np.nan_to_num)` returned a new array that was never assigned back.
- **#3608** — Signal left permanently unfolded after lazy SVD: `self._unfolded4decomposition is False` (identity test, not assignment).
- **#3609** — Navigation mask applied with wrong element order for signals with a multi-dimensional navigation space.
- **#3610** — `_block_iterator` only read the first signal chunk when the on-disk chunk size was smaller than the full signal length (sub-signal HDF5 chunking).
- **#3611** — `ORNMF` hung indefinitely on data with a negative mean (`np.sqrt(negative)` → `NaN` → convergence check never triggers).
- **#3612** — Lazy decomposition: NaN-fill after `reproject` used a fragile string comparison instead of explicit boolean flags; could corrupt factor/loading shapes for ORPCA/ORNMF.
- **#3615** — `decomposition()`: `navigation_size < 2` raised `AttributeError` instead of `ValueError` on the lazy path; no validation for `output_dimension <= 0` or non-integer values.
- **#3616** — `get_bss_model()` corrupted `learning_results` by overwriting `lr.factors`/`lr.loadings` with `bss_factors`/`bss_loadings` in-place.

---

### `svd_solver` option for lazy SVD
Three solvers are now available via the `svd_solver` parameter on `LazySignal.decomposition()`:

- `'randomized'` (default) — randomised truncated SVD via `da.linalg.svd_compressed`; requires `output_dimension`; factors/loadings are numpy arrays.
- `'incremental'` — out-of-core mini-batch streaming via `ISVD` (`sklearn.decomposition.IncrementalPCA` subclass); scales to datasets larger than RAM; factors/loadings are numpy arrays.
- `'full'` — exact SVD via `da.linalg.svd` (TSQR; uses the same backend as HyperSpy prior to v2.5); factors/loadings are **lazy dask arrays**; enables a fully lazy pipeline (see below).

### Fully lazy pipeline
With `svd_solver='full'`, the entire pipeline from decomposition through model reconstruction to saving can be kept lazy — nothing is materialised until `.save()` is called:

```python
s.decomposition(algorithm="SVD", svd_solver="full", output_dimension=3)
model = s.get_decomposition_model()   # lazy LazySignal
model.save("model.hspy")              # streams chunk by chunk
```

The `reproject` steps are also lazy for `svd_solver='full'`: navigation/signal reproject use dask matmuls that stream over chunks. Only the *produced* result (small `nav×k` loadings or `sig×k` factors) is computed eagerly; the unrequested array stays lazy.

### `lazy` and `chunks` keywords on `get_decomposition_model()` and `get_bss_model()`
New keyword arguments on both methods (base class, all signal types):

- `lazy=None` (default): auto-detect — lazy if the signal is lazy, eager otherwise.
- `lazy=True`: always return a `LazySignal`; numpy factors/loadings are wrapped with `da.from_array` locally without mutating `learning_results`.
- `lazy=False`: always return an eager signal, computing immediately.
- `chunks`: chunk shape passed to `da.from_array` when wrapping numpy arrays as dask (only applies when `lazy=True` or the signal is lazy).

### Other new features
- Add `centre` parameter (`"navigation"`, `"signal"`, `None`) to lazy SVD decomposition (all three solvers), consistent with the non-lazy interface.
- Navigation and signal masks are now correctly respected during lazy SVD.
- Add `LazySignal.normalize_poissonian_noise()` as a standalone method, available independently of decomposition.
- Implement `reproject='signal'` and `reproject='both'` for lazy SVD and PCA.
- Add `algorithm='NMF'` for lazy signals via `sklearn.decomposition.MiniBatchNMF` (scikit-learn ≥ 1.1).
- Accept any custom sklearn-like estimator as `algorithm`; objects with `partial_fit` are used incrementally.
- Add `svd_solver` parameter for API parity with non-lazy decomposition. The `auto_transpose` parameter is accepted for API compatibility but is a no-op (benchmarking showed transposition is never advantageous for the incremental solver).

---

## Refactoring

- **Incremental SVD unified with other streaming algorithms** — `svd_solver='incremental'` no longer has a standalone `unfold()`/`rechunk()`/`fold()` block. It now uses the same `_block_iterator`-based path as `"PCA"`, `"NMF"`, `"ORPCA"`, and `"ORNMF"`, removing ~40 lines of duplicated infrastructure.
- **Centralised `output_dimension` and `svd_solver` validation** — the checks that `output_dimension` is a positive integer and that it is supplied for solvers that require it (`'randomized'`, `'incremental'`) have been moved into `_validate_decomposition_inputs()` (shared by both lazy and non-lazy paths), so all constraints live in one place.

---

## Deprecations

- **#3613** — `ORPCA.fit()`, `ORPCA.project()`, `ORPCA.finish()` and the `ORNMF` equivalents are deprecated in favour of the new sklearn-compatible API: `partial_fit()`, `transform()`, `components_`.

---

## Documentation

- `doc/user_guide/big_data.rst`: updated lazy algorithm table; documented `centre`/`svd_solver`, NMF, custom estimators, Poissonian noise normalisation, the fully lazy pipeline, and the `lazy`/`chunks` kwargs on `get_decomposition_model()`. Array-type table expanded to show per-`reproject`-mode laziness for `svd_solver='full'`. Added guidance note comparing all algorithms by memory footprint and speed; clarified that `svd_solver='incremental'` has the lowest peak memory among the three SVD solvers (not overall, since `"PCA"` and `"NMF"` use less).
- `doc/user_guide/mva/decomposition.rst`: added lazy-NMF note and deprecation notes for ORPCA/ORNMF; removed stale restriction on `centre` for lazy SVD; added masks/reprojection and model-laziness sections.
- `doc/user_guide/mva/bss.rst`: added note on the new `lazy` kwarg for `get_bss_model()`.
- Docstrings on `LazySignal.decomposition`, `_calculate_recmatrix`, `get_decomposition_model`, and `get_bss_model` updated with Notes sections, array-type tables, parameter descriptions, and examples.
- Fixed 13 docstring and User Guide wording issues identified in review: broken Sphinx cross-references, self-referential See Also, misleading one-liner, incomplete algorithm lists, and minor style inconsistencies across `lazy.py`, `_mva.py`, `incremental_svd.py`, `big_data.rst`, and `decomposition.rst`.

### CODE REVIEW

**Comment by CSSFrancis** [comment]
@francisco-dlp This is a lot to review... This is definitely something that needs to be done, but is there a better way of doing this? 83 commits are a lot to go through, and I'd rather not just say LGTM. 

Maybe going through in person might help. Otherwise, it might be possible to just say -- "Divide this into separately scoped PRs and create a .md file which walks through the changes and motivations for each.  Create one overarching .md file which describes the full architecture change, includes important elements from each and motives the changes"

I've found that AI generated code usually does a bit better on the second pass.  Things that work well for humans work well for things trained to act like humans.  Even something like creating excessive documentation on changes help to force good practices and will often times end up flagging things.  AI has limited context. It's a lot more like a scapple than an axe.  It's good at doing specific things but struggles with larger contexts.

**Comment by ericpre** [comment]
I agree with @CSSFrancis, there is too many changes across too many commits for this PR to be reviewed by an human. @francisco-dlp, can it be split into smaller PRs?

**Comment by francisco-dlp** [comment]
Thanks for your comments @CSSFrancis and @ericpre. The huge number of commits (and code to review) is the outcome of asking the AI agent to aim for parity with the non-lazy equivalent features. The idea is that the AI should perform best at the task by working toward this very well defined goal, and that the existing equivalent code for eager signals (including tests), should help it stay on track.

I can try to split this into separate PRs but, for the AI agent, it is better to work within this branch that has an unified clear objective. Also, split or not, this is a lot of code to review for a human. In other words: we are the bottleneck here--should we adapt our reviewing methods to the age of AI for situations such as this one?

**Comment by CSSFrancis** [comment]
@francisco-dlp I think we do have to think about reviewing this a little differently.  I agree that sometimes it's better to create a single large change.  I like using the claude [superpowers](https://github.com/obra/superpowers) as it's good at actually following good software development.  I tend to go through about 10-20 different planning steps before implementing anything.  Usually I'll have an idea, go back and forth for a day or two and then try implementing it. 

For this PR, I'd maybe think about a prompt like:

```make a plan for splitting this PR in multiple self contained PR's each with a limited scope and a self contained defined objective. The total functionality should be maintained but each PR's should be limited in scope with defined tests and be optimized for human-review. ```

After that:

1. I tend to do a pretty in depth self review.  My steps are:
    a. Ask the AI agent to review it's self (usually I do this ~2 times) for larger changes.  Often times these reviews aren't great but they do sometimes catch things.
    b. Ask gituhb copilot to review.  This usually works much better.  It tends to be a little better at catching inconsistencies. 
    c. Review the code myself.  I like to do this in GitHub kind of like I was reviewing someone elses code. 

2. Then put it out to other people to review. 


In all of these cases the human is the weakest link, but also the most important.  From a quick look there are lots of deprecations but they are not using the [deprecated](https://github.com/hyperspy/hyperspy/blob/6b303ec4be48a446233d75c5605de319d2cf8b2d/hyperspy/decorators.py#L121) decorator. In addition there is no version at which they are going to be removed.

**Comment by francisco-dlp** [comment]
Closed in favor of the restructured implementation across 5 focused, separately reviewable PRs as requested by reviewers (@CSSFrancis, @ericpre):

- **#3650**: Bugfixes for lazy decomposition (issues #3607-#3612) — target v2.4.1
- **#3651**: sklearn-compatible API for ORPCA/ORNMF with @deprecated decorator — target v2.5.0
- **#3652**: ISVD — Incremental SVD class with centering disabled — target v2.5.0
- **#3653**: Lazy SVD pipeline (svd_solver, centre, reproject, NMF, custom estimators) — target v2.5.0
- **#3654**: Documentation updates for lazy decomposition overhaul — target v2.5.0

All reviewer feedback addressed:
- ✅ Divided into separately scoped PRs (5 focused branches)
- ✅ Deprecations use @deprecated decorator with removal version (since="2.5", removal="3.0")
- ✅ 7 Copilot-identified axis-order bugs fixed
- ✅ Plain backticks in all changelogs (no :meth:, :class:, :attr:)
- ✅ ALL-DIFFERENT array dimensions in tests
- ✅ 854 tests passing, ruff clean

The 5 PRs form a sequential dependency chain (#3650 → #3651 → #3652 → #3653 → #3654) and should be reviewed in order.

Assisted-by: opencode:deepseek-v4-pro
